### ISSUE TITLE
Syntax error on shebang line in PsychoPy (Python 3) when compiling code from Builder

### ISSUE THREAD

**Issue by dvbridges** [psychopy/psychopy#2235]
When compiling Python code from Builder using PsychoPy running on Python 3 (on Windows 10), the shebang line contains a syntax error:
```
ï»¿#!/usr/bin/env python
# -*- coding: utf-8 -*-
```

**Comment by hoechenberger** [comment]
Could you please attach a file like this? This could be the BOM we're adding now by default in many places – don't remember if we modified Builder output too? When you say "contains a syntax error", do you mean this is how it gets displayed in Coder after compiling? Do you see this symbol in a "proper" editor too? (I suspect not) – Could mean we'd need to adjust the file-loading mechanism in Coder (to use `utf-8-sig` encoding)

**Comment by hoechenberger** [comment]
So yeah I'm very certain it's a BOM, see https://en.wikipedia.org/wiki/Byte_order_mark#Byte_order_marks_by_encoding ("Bytes as CP1252 characters" column for UTF-8, i.e., first row)

**Comment by hoechenberger** [comment]
I suspect we should make use of wxPython's [`StyledTextCtrl.StyleSetFontEncoding()`](https://wxpython.org/Phoenix/docs/html/wx.stc.StyledTextCtrl.html#wx.stc.StyledTextCtrl.StyleSetFontEncoding)

**Comment by dvbridges** [comment]
Thanks Richard, I cannot run the file from Coder or from the Python interpreter, it throws the ```Syntax Error: invalid character in identifier```, however Builder itself outputs the code correctly, so I guess we see why the outputs two differ.

**Comment by peircej** [comment]
@hoechenberger maybe it's just csv files that need the BOM adding?

**Comment by hoechenberger** [comment]
> What I mean is, the exp_lastrun.py script is ok.

Was already very surprised since I first read your comment as this was not the case. For reference, here's a small test case:
```python
import pathlib


def create_file(*, filename, content):
    with open(filename, 'w', encoding='utf-8-sig') as f:
        f.write(content)


filename = pathlib.Path('/tmp/foo.py')
content = """#!/usr/bin/env python
# -*- coding: utf-8 -*-

print('Hello there.')
"""

create_file(filename=filename, content=content)
```

The resulting file has a BOM and can be run without any issues.
```
(base) [hoechenberger:/tmp]$ python foo.py
Hello there.
(base) [hoechenberger:/tmp]$
```

> @hoechenberger maybe it's just csv files that need the BOM adding?

Actually, as far as I can see, no: if we want to support e.g. Chinese string literals in the scripts, they need to be saved as `utf-8-sig` as well.

Since `exp_lastrun.py` seems to be alright – which should contain the exact same code as what is displayed in the Coder, right? –, I think we should try setting the encoding for the `StyleTextCtrl` (as proposed here: https://github.com/psychopy/psychopy/issues/2235#issuecomment-458229997), since by default WX uses the system default encoding, which, on your UK or US Windows system, would typically be non-UTF-8.

**Comment by hoechenberger** [comment]
@dvbridges If you open `exp_lastrun.py` in Coder, do you see the same issue?

**Comment by dvbridges** [comment]
No, the ```exp_lastrun.py``` has no issues in Coder - so probably specific to how ```psyexpCompile``` compiles the script.

**Comment by hoechenberger** [comment]
Both [`builder.BuilderFrame.runFile()`](https://github.com/psychopy/psychopy/blob/64449dd68b7e5c0d01d9f893c333821edebe1289/psychopy/app/builder/builder.py#L2125) and [`builder.BuilderFrame.compileScript`](https://github.com/psychopy/psychopy/blob/64449dd68b7e5c0d01d9f893c333821edebe1289/psychopy/app/builder/builder.py#L2311) call [`builder.BuilderFrame.generateScript()`](https://github.com/psychopy/psychopy/blob/64449dd68b7e5c0d01d9f893c333821edebe1289/psychopy/app/builder/builder.py#L2316), which calls `psychopy.scripts.psyexpCompile()`. Should produce the exact same output!

**Comment by hoechenberger** [comment]
> I think we should try setting the encoding for the `StyleTextCtrl` (as proposed here: [#2235 (comment)](https://github.com/psychopy/psychopy/issues/2235#issuecomment-458229997)), since by default WX uses the system default encoding, which, on your UK or US Windows system, would typically be non-UTF-8, meaning the BOM won't be handled correctly.

Then again, we open and read the file with the correct encoding before passing the contents to WX, so the BOM should actually have been removed by then, no?

**Comment by dvbridges** [comment]
Yes, ok how about this line - as it is specific to Python 3, and this is where the file is read after being compiled - open the file with utf-8-sig encoding

https://github.com/psychopy/psychopy/blob/64449dd68b7e5c0d01d9f893c333821edebe1289/psychopy/app/coder/coder.py#L2138

**Comment by dvbridges** [comment]
Yes, it works :). Good idea using the same the context manager.

**Comment by hoechenberger** [comment]
Great! Can you submit a PR?

**Comment by dvbridges** [comment]
Yep, no problem

**Comment by peircej** [comment]
> Actually, as far as I can see, no: if we want to support e.g. Chinese string literals in the scripts, they need to be saved as utf-8-sig as well.

Hmm, what about the script encoding line itself:
`# -*- coding: utf-8 -*-`
Should that then be
`# -*- coding: utf-8-sig -*-`
?

**Comment by hoechenberger** [comment]
> Hmm, what about the script encoding line itself:

I thought about that too, but I concluded that, since the BOM header must have been correctly "stripped" from the sources for the Python interpreter to even reach this line (otherwise – see error in this issue), `utf-8` should be fine here.

And now, thinking even more about this entire issue, maybe you're right – maybe we need `utf-8-sig` for CSV files only? The reason we switched to this charset was bc Excel requires it... Adding a BOM to `utf-8` files is actually uncommon / non-standard, it's just an MS Excel thing. Mmmh.
